tests/simple/testdata/type_deduction.textproto

```diff
@@ -259,3 +259,497 @@ section {
+      deduced_type {
+        list_type {
+          elem_type {
+            abstract_type {
+              name: "optional_type"
+              parameter_types { dyn {} }
+            }
+          }
+        }
+      }
+    }
+  }
+  test {
+    name: "optional_in_ternary"
+    expr: "true ? optional.of(dyn(1)) : optional.of(1)"
+    check_only: true
+    typed_result {
+      deduced_type {
+        abstract_type {
+          name: "optional_type"
+          parameter_types { dyn {} }
+        }
+      }
+    }
+  }
+}
+
+# Wrapper types have some special handling. They generally behave like a union
+# type e.g. `null | int64` with auto-unboxing, but there are some quirks.
+section {
+  name: "wrappers"
+  test {
+    name: "wrapper_promotion"
+    type_env {
+      name: "msg"
+      ident {
+        type {
+          message_type: "cel.expr.conformance.proto3.TestAllTypes"
+        }
+      }
+    }
+    expr: "[msg.single_int64_wrapper, msg.single_int64]"
+    check_only: true
+    typed_result {
+      deduced_type {
+        list_type {
+          elem_type {
+            wrapper: INT64
+          }
+        }
+      }
+    }
+  }
+  test {
+    name: "wrapper_promotion_2"
+    type_env {
+      name: "msg"
+      ident {
+        type {
+          message_type: "cel.expr.conformance.proto3.TestAllTypes"
+        }
+      }
+    }
+    expr: "[msg.single_int64, msg.single_int64_wrapper]"
+    check_only: true
+    typed_result {
+      deduced_type {
+        list_type {
+          elem_type {
+            wrapper: INT64
+          }
+        }
+      }
+    }
+  }
+  test {
+    name: "wrapper_dyn_promotion"
+    type_env {
+      name: "msg"
+      ident {
+        type {
+          message_type: "cel.expr.conformance.proto3.TestAllTypes"
+        }
+      }
+    }
+    expr: "[msg.single_int64_wrapper, msg.single_int64, dyn(1)]"
+    check_only: true
+    typed_result {
+      deduced_type {
+        list_type {
+          elem_type {
+            dyn {}
+          }
+        }
+      }
+    }
+  }
+  test {
+    name: "wrapper_dyn_promotion_2"
+    type_env {
+      name: "msg"
+      ident {
+        type {
+          message_type: "cel.expr.conformance.proto3.TestAllTypes"
+        }
+      }
+    }
+    expr: "[dyn(1), msg.single_int64_wrapper, msg.single_int64]"
+    check_only: true
+    typed_result {
+      deduced_type {
+        list_type {
+          elem_type {
+            dyn {}
+          }
+        }
+      }
+    }
+  }
+  test {
+    name: "wrapper_dyn_promotion_2"
+    type_env {
+      name: "msg"
+      ident {
+        type {
+          message_type: "cel.expr.conformance.proto3.TestAllTypes"
+        }
+      }
+    }
+    expr: "[dyn(1), msg.single_int64_wrapper, msg.single_int64]"
+    check_only: true
+    typed_result {
+      deduced_type {
+        list_type {
+          elem_type {
+            dyn {}
+          }
+        }
+      }
+    }
+  }
+}
+
+section {
+  name: "type_parameters"
```

Comment:
Could you test the following expressions as well:

`tuple(1, 2, 3) == tuple(1, dyn(2u), dyn(3.0))`
`tuple(1, dyn(2u), dyn(3.0)) == tuple(1, 2, 3)`

Thanks!

Reply:
Done